File: lib/core/services/group_chat_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../../features/chat/models/message_model.dart';

class GroupChatService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  static const String _collection = 'groupChats';

  /// Subscribe to messages for a given group (real-time)
  /// Returns a Stream of messages
  Stream<List<MessageModel>> subscribeToMessages(String groupId) {
    return _firestore
        .collection(_collection)
        .doc(groupId)
        .collection('messages')
        .orderBy('createdAt', descending: false)
        .limit(500)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) => MessageModel.fromFirestore(doc)).toList();
    });
  }

  /// Send a message
  Future<void> sendMessage({
    required String groupId,
    required String senderId,
    required String senderName,
    required String text,
  }) async {
    if (text.trim().isEmpty) return;

    final message = MessageModel(
      id: '',
      senderId: senderId,
      senderName: senderName,
      text: text.trim(),
      createdAt: DateTime.now(),
    );

    await _firestore
        .collection(_collection)
        .doc(groupId)
        .collection('messages')
        .add(message.toFirestore());
  }
}

File: web-react/src/core/services/GroupChatService.js
import {
  collection,
  addDoc,
  query,
  orderBy,
  limit,
  serverTimestamp,
  onSnapshot
} from 'firebase/firestore';
import { db } from '../../lib/firebase';

const COLLECTION = 'groupChats';

export const GroupChatService = {
  /**
   * Subscribe to messages for a given group (real-time)
   * @param {string} groupId
   * @param {(msgs: Array) => void} callback
   * @returns {() => void} unsubscribe function
   */
  subscribe(groupId, callback) {
    const q = query(
      collection(db, COLLECTION, groupId, 'messages'),
      orderBy('createdAt', 'asc'),
      limit(500)
    );

    return onSnapshot(q, (snap) => {
      const msgs = snap.docs.map((d) => ({ id: d.id, ...d.data() }));
      callback(msgs);
    });
  },

  /**
   * Send a message
   * @param {string} groupId
   * @param {string} senderId
   * @param {string} senderName
   * @param {string} text
   */
  async sendMessage(groupId, senderId, senderName, text) {
    if (!text?.trim()) return;
    await addDoc(collection(db, COLLECTION, groupId, 'messages'), {
      senderId,
      senderName,
      text: text.trim(),
      createdAt: serverTimestamp()
    });
  }
};